### PR TITLE
As an item, I want my estimated next purchase date to be computed at the time my purchase is recorded in the database so the app can learn how often I buy different items.

### LINKED ISSUE #10: 10. As an item, I want my estimated next purchase date to be computed at the time my purchase is recorded in the database so the app can learn how often I buy different items.
AC:

* When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database

The script `calculateEstimate` from `@the-collab-lab/shopping-list-utils` should be used to calculate the next estimated purchase interval. This should already be installed in your project from [issue 3](https://github.com/the-collab-lab/tcl-33-smart-shopping-list/issues/3)

Notes:

* Think through the data that is needed as inputs to the function in `estimates.js` and how to use the output.
* It is likely at this point you'll need to adjust the schema for an item to accommodate the data points required.

### PR DESCRIPTION
**Pull request by aidantorrence**
## Description

We added functionality that calculates an estimate of the number of days for the next purchase using the calculateEstimate utility and records it to the firestore. 

## Related Issue

closes #10 

## Acceptance Criteria

- [ ] When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### After

image

### CODE REVIEW

**Comment by andreamartz** [comment]
Thanks for working on this. 
I created a new list, put 'carrots' on the list, and marked it as purchased.
Then, I looked in the database and saw this:
carrots from list with token 'cat fine valve'

File: src/List.js [hunk 1]
```diff
@@ -66,13 +86,33 @@ export function List() {
...
     const checked = event.target.checked;
     if (checked) {
       const itemRef = doc(db, 'shopping-list', id);
-      setDoc(itemRef, { lastPurchasedDate: date.getTime() }, { merge: true });
+      setDoc(
+        itemRef,
+        {
+          lastPurchasedDate: date.getTime(),
+          estimatedPurchaseDate: calculateEstimate(
+            item.estimatedPurchaseDate,
```
Comment (gabynr) [inline src/List.js:101]: I'm not sure, but I just have asked  in the slack, but I think, here you need to send the values in the attribute `days`  that select in the radio button

**Comment by gabynr** [comment]
I'm not sure, but I just have asked in the slack, but I think, in  ` calculateEstimate` here 
            `estimatedPurchaseDate`, you need to send the values of the attribute days that select in the radio button.

File: src/List.js [hunk 2]
```diff
@@ -66,13 +86,33 @@ export function List() {
...
+          totalPurchases: item.totalPurchases + 1,
+        },
+        { merge: true },
+      );
     } else {
       const itemRef = doc(db, 'shopping-list', id);
-      setDoc(itemRef, { lastPurchasedDate: null }, { merge: true });
+      setDoc(
+        itemRef,
+        { lastPurchasedDate: null, totalPurchases: item.totalPurchases - 1 },
```
Comment (gabynr) [inline src/List.js:113]: if the user click unchecked and `lastPurchasedDate` changes to  `null`  we'll lose the information from  the `lastPurchasedDate` before it was unchecked?

**Review by gabynr — CHANGES_REQUESTED** [review]
I have a couple of questions for you

**Review by andreamartz — CHANGES_REQUESTED** [review]

File: src/List.js (hunk 2 above)
Comment (MikeBLambert) [reply src/List.js:113]: One thing to consider is whether a user should be allowed to uncheck? In the past groups have made the decision to not allow users to uncheck the checkbox.

File: src/List.js [hunk 3]
```diff
@@ -66,13 +86,33 @@ export function List() {
 
   const handleChange = async (id, event) => {
     let date = new Date();
+    const item = items.find((element) => element.id === id);
+    const daysSinceLastTransaction = item.lastPurchasedDate
+      ? Math.round((new Date() - item.lastPurchasedDate) / 1000 / 60 / 60 / 24)
```
Comment (MikeBLambert) [inline src/List.js:91]: Maybe a variable here would help to make it more clear what `Math.round((new Date() - item.lastPurchasedDate) / 1000 / 60 / 60 / 24)` is.

File: src/List.js [hunk 4]
```diff
@@ -66,13 +86,33 @@ export function List() {
 
   const handleChange = async (id, event) => {
     let date = new Date();
+    const item = items.find((element) => element.id === id);
+    const daysSinceLastTransaction = item.lastPurchasedDate
```
Comment (MikeBLambert) [inline src/List.js:90]: `.find` can return undefined, meaning that `item` could be undefined. If that was the case then we would get a null pointer error here. We can avoid this by doing `item?.lastPurchasedDate`.

File: src/List.js (hunk 3 above)
Comment (bajancode) [reply src/List.js:91]: Thanks Mike, done!

File: src/List.js [hunk 5]
```diff
@@ -33,11 +34,30 @@ export function List() {
...
+                id,
+                name,
+                userToken,
+                lastPurchasedDate,
+                estimatedPurchaseDate,
+                totalPurchases,
+              },
+            ];
+          },
+          [token],
```
[the remark this answers is not in the document]
Comment (bajancode) [reply src/List.js:59]: Great points thank you! Taken care of!

File: src/List.js (hunk 1 above)
Comment (bajancode) [reply src/List.js:101]: Thank you, we've incorporated this suggestion!

File: src/List.js (hunk 2 above)
Comment (bajancode) [reply src/List.js:113]: Thank you, we removed the ability to uncheck items so that we're in keeping with all the other groups!

**Review by gabynr — COMMENTED** [review]
🚢 Thank you for accepting my suggestions. Great Work!!! 😊

**Review by andreamartz — APPROVED** [review]

**Review by abidishajia — APPROVED** [review]

**Pull request merged**